codegen_llvm_back: improve allocations

ljedrz opened the pull request:
This commit was split out from https://github.com/rust-lang/rust/pull/54864. Last time it was causing an LLVM OOM, which was most probably caused by not collecting the globals.

- preallocate vectors of known length
- `extend` instead of `append` where the argument is consumable
- turn 2 `push` loops into `extend`s
- create a vector from a function producing one instead of using `extend_from_slice` on it
- consume `modules` when no longer needed
- ~~return an `impl Iterator` from `generate_lto_work`~~
- ~~don't `collect` `globals`, as they are iterated over and consumed right afterwards~~

While I'm hoping it won't cause an OOM anymore, I would still consider this a "high-risk" PR and not roll it up.

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

nagisa commented:
@bors try

Lets do a perf run :)

bors commented:
:hourglass: Trying commit dc1b2c75dbe42dff700087410475aedf3450dbe4 with merge 684fb37161c61f3e817fb2dcf81cd7c9097f3936...

bors commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/rust-lang/rust/builds/453607875?utm_source=github_status&utm_medium=notification)
State: approved= try=True

rust-timer commented:
Finished benchmarking try commit 684fb37161c61f3e817fb2dcf81cd7c9097f3936

nagisa commented:
It seems that both instruction counts and [max-rss have suffered](https://perf.rust-lang.org/compare.html?start=b76ee83254ec0398da554f25c2168d917ba60f1c&end=684fb37161c61f3e817fb2dcf81cd7c9097f3936&stat=max-rss) a fair hit, while for most benchmarks the minimum rss has also dropped significantly. Essentially this means we have increased deviation, without a clear win in mean rss.

nagisa commented:
@bors try

bors commented:
:hourglass: Trying commit 9e8bafc24b60008353f5f4e8027379a10bc7bb35 with merge f7360e5b2e5b2ed5e1696af257b365e3cc69981d...

nagisa commented:
@rust-timer build f7360e5b2e5b2ed5e1696af257b365e3cc69981d

rust-timer commented:
Success: Queued f7360e5b2e5b2ed5e1696af257b365e3cc69981d with parent 0195812aeafeecaa8760a4ddceae187472db8fe6, [comparison URL](https://perf.rust-lang.org/compare.html?start=0195812aeafeecaa8760a4ddceae187472db8fe6&end=f7360e5b2e5b2ed5e1696af257b365e3cc69981d).

rust-timer commented:
Finished benchmarking try commit f7360e5b2e5b2ed5e1696af257b365e3cc69981d

ljedrz commented:
Since the changes at this point are all pretty harmless I'd say that the benchmark results are statistical noise. 

That being said, these changes don't seem to be beneficial performance-wise, so I'm ok with closing the PR, unless you believe that they are a readability improvement / more idiomatic.

nagisa commented:
Uhh, no matter how I look at it, the [max-rss results](https://perf.rust-lang.org/compare.html?start=0195812aeafeecaa8760a4ddceae187472db8fe6&end=f7360e5b2e5b2ed5e1696af257b365e3cc69981d&stat=max-rss) still seem like a hit-or-miss.

nikic commented:
@nagisa max-rss has very high variance. Here's how it looks for a random recent commit: http://perf.rust-lang.org/compare.html?start=ca79ecd6940e30d4b2466bf378632efcdf5745c7&end=775eab58835f9bc0f1f01ccbb79725dce0c73b51&stat=max-rss The results for this PR seem to be within the "usual" noise.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit 9e8bafc24b60008353f5f4e8027379a10bc7bb35 has been approved by `nagisa`

<!-- @bors r=nagisa 9e8bafc24b60008353f5f4e8027379a10bc7bb35 -->

pietroalbini commented:
@bors r-

Still OOMing on AppVeyor.

nagisa commented:
Could it be that with this patch we overcommit virtual memory that ends up
being never used? On Windows overcommit is not possible which would explain
oom there but no observable regressions on perf runs?

On Fri, Nov 16, 2018, 11:07 Pietro Albini <notifications@github.com wrote:

> @bors <https://github.com/bors> r-
>
> Still OOMing on AppVeyor.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/55871#issuecomment-439328019>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApc0jXIny6RIi09_7OWVgMoCEYUM2Lcks5uvoA6gaJpZM4YYf06>
> .
>

ljedrz commented:
Maybe, but how? These changes shouldn't negatively impact allocations - IMO they should make them easier to optimize.

nagisa commented:
> IMO they should make them easier to optimize.

The optimizability is irrelevant if it is indeed what I think it is. And is hardly related to the number of allocations, but rather to the size of them. Alas, rust perf does not collect the information of interest to tell for sure.

It would be interesting to do a perf run (and they really should be run that way all the time) with overcommit disabled.

ljedrz commented:
> hardly related to the number of allocations

As long as the length of the iterator is known, changing a `push` loop to an `extend` reduces the number of allocations to just one.

bors commented:
:umbrella: The latest upstream changes (presumably #55627) made this pull request unmergeable. Please resolve the merge conflicts.

nagisa commented:
I’m not sure we’re talking about the same thing. Even though the following snippet would OOM on windows, it would work just fine on UNIXes due to overcommit:

```
auto x = new uint8_t[2 << 42]; // 4TiB or thereabouts…
x[0] = 42;
```

This has nothing to do with known length or allocation count.

@rust-lang/infra is it possible to make @rust-timer to collect additional information (e.g. max-virtual-mem, in addition to max-rss…)? What repository should I fill an issue to request this?

---

While something like this happening is clearly a bug *somewhere* (and I cannot tell where exactly), it is fairly obvious some change from those in the commit are making LLVM to commit too much memory that likely ends up never being used.

One thing to debug this you could do is to compile stage1 core on your own UNIX machine and see what the maximum virtual memory ends up being. If it ends up being significantly larger than RSS, that would confirm my suspicions.

Another thing you could do is disabling `overcommit` (if you’re on Linux, that’s `sysctl vm.overcommit=2`).

Finally, we could also just bisect – there aren’t that many different changes in this PR. We could try landing them one by one (though there’s a danger that all these changes are cumulatively slightly raising the committed memory and none of them would fail CI on their own).

pietroalbini commented:
> is it possible to make @rust-timer to collect additional information (e.g. max-virtual-mem, in addition to max-rss…)? What repository should I fill an issue to request this?

https://github.com/rust-lang-nursery/rustc-perf

rust-timer commented:
Insufficient permissions to issue commands to rust-timer.

ljedrz commented:
@nagisa ah, ok, thanks for the explanation; I wasn't thinking in terms of a possible memory bug.

I will do some test builds on a Linux machine when I have a bit of free time.

nikic commented on src/librustc_codegen_llvm/back/write.rs:
```diff
@@ -2586,8 +2586,8 @@ fn create_msvc_imps(cgcx: &CodegenContext, llcx: &llvm::Context, llmod: &llvm::M
-            })
-            .collect::<Vec<_>>();
+            });
+
```
You're iterating over globals here and then adding new globals in the loop below. With the collect that's fine, as you'll only iterate existing globals. Without the collect this is going to be an infinite loop and you OOM.

ljedrz replied:
Weird that this doesn't seem to cause issues on architectures different than `i686` (at least that's the one the OOM was being hit on appveyor); I can compile with these changes on `x64` without issues.

ljedrz commented:
@nikomatsakis At this point I was interested more in the possibility of uncovering some memory-related bug (as described by @nagisa). I think @nikic is onto something.

ljedrz commented:
@nikomatsakis I just remembered that the initial version of this PR included a change that had a [-2.5% win for style-servo-opt](https://perf.rust-lang.org/compare.html?start=b76ee83254ec0398da554f25c2168d917ba60f1c&end=684fb37161c61f3e817fb2dcf81cd7c9097f3936) which is considerable, especially since it is a huge benchmark; I rebased, re-included the win for servo and removed the problematic bit that @nikic marked as the one causing the OOM, so hopefully now it's good to go :crossed_fingers:.

nagisa commented:
Let's see if bors can be controlled over mail

@bors r+

On Mon, Dec 3, 2018, 11:39 ljedrz <notifications@github.com wrote:

> @nikomatsakis <https://github.com/nikomatsakis> I just remembered that
> the initial version of this PR included a change that had a -2.5% win for
> style-servo-opt
> <https://perf.rust-lang.org/compare.html?start=b76ee83254ec0398da554f25c2168d917ba60f1c&end=684fb37161c61f3e817fb2dcf81cd7c9097f3936>
> which is considerable, especially since it is a huge benchmark; I rebased,
> re-included the win for servo and removed the problematic bit that @nikic
> <https://github.com/nikic> marked as the one causing the OOM, so
> hopefully now it's good to go 🤞.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/55871#issuecomment-443646902>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApc0lvHJjIT9JVeBCjiPkaEmibS8ON6ks5u1PFkgaJpZM4YYf06>
> .
>

bors commented:
:pushpin: Commit 2043d30c2e8c912dd04be2913528979a1977c330 has been approved by `nagisa`

<!-- @bors r=nagisa 2043d30c2e8c912dd04be2913528979a1977c330 -->

bors commented:
:hourglass: Testing commit 2043d30c2e8c912dd04be2913528979a1977c330 with merge 431e0ab62f7730f11db693c23f48403e4c719f82...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/20756539), [status-travis](https://travis-ci.org/rust-lang/rust/builds/463304294?utm_source=github_status&utm_medium=notification)
Approved by: nagisa
Pushing 431e0ab62f7730f11db693c23f48403e4c719f82 to master...
